aero_vloc/localization_pipeline/localization_pipeline.py

```diff
@@ -0,0 +1,173 @@
+        self,
+        matched_kpts_query,
+        matched_kpts_reference,
+        sat_image,
+        drone_img,
+        new_shape,
+    ):
+        drone_img = cv2.resize(cv2.imread(str(drone_img.path)), new_shape)
+        sat_img = cv2.resize(
+            cv2.imread(str(sat_image.path)),
+            (self.feature_matcher.resize, self.feature_matcher.resize),
+        )
+        matches = [cv2.DMatch(i, i, 1) for i in range(len(matched_kpts_query))]
+        matched_kpts_query = [cv2.KeyPoint(x, y, 1) for x, y in matched_kpts_query]
+        matched_kpts_reference = [
+            cv2.KeyPoint(x, y, 1) for x, y in matched_kpts_reference
+        ]
+        img = cv2.drawMatches(
+            drone_img,
+            matched_kpts_query,
+            sat_img,
+            matched_kpts_reference,
+            matches,
+            None,
+        )
+        return img
+
+    def localize(
+        self, query_image: UAVImage, k_closest: int, visualize: bool = False
+    ) -> Optional[Tuple[float, float, Optional[np.ndarray]]]:
+        """
+        Calculates UAV location using global localization and homography
+        :param query_image: The image for which the location should be calculated
+        :param k_closest: Specifies how many predictions for each query the global localization should make.
+        If this value is greater than 1, the best match will be chosen with local matcher
+        :param visualize: Determines if visualization of keypoint matching
+        between the query keypoints and the selected satellite image should be returned
+        :return: Latitude, longitude, visualization (optional). None, if localization failed
+        """
+        retrieve_result = self.__retrieve_image(query_image, k_closest)
+        if retrieve_result is None:
+            return None
+        res_prediction, matched_kpts_query, matched_kpts_reference = retrieve_result
+        if len(matched_kpts_reference) < 4:
+            print("Not enough points for homography")
+            return None
+        chosen_sat_image = self.sat_map.tiles[res_prediction]
+
+        h_new, w_new = get_new_size(
+            *cv2.imread(str(query_image.path)).shape[:2], self.feature_matcher.resize
+        )
+        M, mask = cv2.findHomography(
+            matched_kpts_query, matched_kpts_reference, cv2.RANSAC, 5.0
+        )
+        pts = np.float32(
+            [[0, 0], [0, h_new - 1], [w_new - 1, h_new - 1], [w_new - 1, 0]]
+        ).reshape(-1, 1, 2)
+        try:
+            dst = cv2.perspectiveTransform(pts, M)
+        except cv2.error:
+            print("Perspective transform error. Abort matching")
+            return None
+
+        moments = cv2.moments(dst)
+        cX = int(moments["m10"] / moments["m00"])
+        cY = int(moments["m01"] / moments["m00"])
+
+        center = (cX / self.feature_matcher.resize, cY / self.feature_matcher.resize)
+
+        latitude = chosen_sat_image.top_left_lat + abs(center[1]) * (
```

Comment:
Geo-referencing to be moved to separate class later, let's add issue
That will be Anna part

Reply:
Also, reprojection error reporting

Reply:
#2